src/app/_blocks/BlogpostChapters/index.tsx

```diff
@@ -1,49 +1,59 @@
 "use client";
 
-import { useEffect, useState } from "react";
+import { JSX, useEffect, useState } from "react";
 
 import styles from "./styles.module.css";
+import { getChapters, sanitizeAndAddChapters } from "@/app/_utilities/sanitizeAndAddChapters";
 
-export default function BlogpostChapters() {
+export default function BlogpostChapters({ content_html }: { content_html: string }): JSX.Element {
   const [visibleChapter, setVisibleChapter] = useState("");
-  const [chapters, setChapters] = useState<string[]>([]);
-  // TODO: Fix chapter navigator
-  useEffect(() => {
-    const extractHeadings = () => {
-      // Query all headings in the DOM (e.g., h1, h2, h3...)
-      const chapterList: string[] = [];
-      const headings = Array.from(document.querySelectorAll("h1, h2, h3"));
-      headings.map(i => chapterList.push(i.innerHTML.trim()));
 
-      setChapters(chapterList);
-    };
+  const chapters = getChapters(content_html);
+useEffect(() => {
+  const headings = document.querySelectorAll('h1, h2, h3, h4, h5, h6');
 
-    extractHeadings();
+  const observer = new IntersectionObserver((entries) => {
+    let visibleId = "";
 
+    entries.forEach((entry) => {
+      if (entry.isIntersecting || entry.boundingClientRect.bottom < 0) {
+        visibleId = entry.target.id;
+      }
+    });
 
-  }, []);
+    setVisibleChapter(visibleId);
+  }, {
+    root: null,
+    threshold: 0.5,
+  });
+
+  headings.forEach((heading) => observer.observe(heading));
+
+  return () => {
+    headings.forEach((heading) => observer.unobserve(heading));
+  };
+}, [chapters]);
```

Comment:
@davelange regarding blogpost chapter nav, IntersectionObserver seems a bit finicky
links are working fine, but the navbar is not working spectacularly. when there is more than 1 id on the viewport" it makes sense, but when there is only one it sometimes work and sometimes doesn't

couldn't find a better way to achieve this result, is there a better way to achieve this result?

demo of the behavior: https://github.com/user-attachments/assets/6c47308a-710a-4103-9a1b-ddc25baf9155